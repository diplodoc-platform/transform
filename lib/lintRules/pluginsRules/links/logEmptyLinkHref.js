const {bold} = require('chalk');
const {LOG_LEVELS} = require('../../../constants');

const ruleName = 'links-empty-href';

const defaultRuleOptions = {
    level: LOG_LEVELS.WARN,
};

function logEmptyLinkHref({ruleOptions, commonOptions}) {
    const {path, log} = commonOptions;
    const {level} = ruleOptions;

    log[level](`Empty link in ${bold(path)}`);
}

logEmptyLinkHref.ruleName = ruleName;
logEmptyLinkHref.ruleOptions = defaultRuleOptions;

module.exports = logEmptyLinkHref;
