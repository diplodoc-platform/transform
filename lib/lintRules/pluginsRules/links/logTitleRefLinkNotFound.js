const {bold} = require('chalk');
const {LOG_LEVELS} = require('../../../constants');

const ruleName = 'title-ref-link-not-found';

const defaultRuleOptions = {
    level: LOG_LEVELS.WARN,
};

function logTitleRefLinkNotFound({ruleOptions, commonOptions}) {
    const {path, href, log} = commonOptions;
    const {level} = ruleOptions;

    log[level](`Title not found: ${bold(href)} in ${bold(path)}. (${ruleName})`);
}

logTitleRefLinkNotFound.ruleName = ruleName;
logTitleRefLinkNotFound.ruleOptions = defaultRuleOptions;

module.exports = logTitleRefLinkNotFound;
