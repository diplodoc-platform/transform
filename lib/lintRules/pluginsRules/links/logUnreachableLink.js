const {bold} = require('chalk');
const {LOG_LEVELS} = require('../../../constants');

const ruleName = 'unreachable-link';

const defaultRuleOptions = {
    level: LOG_LEVELS.ERROR,
};

function logUnreachableLink({ruleOptions, commonOptions}) {
    const {path, log, href} = commonOptions;
    const {level} = ruleOptions;

    log[level](`Link is unreachable: ${bold(href)} in ${bold(path)}. (${ruleName})`);
}

logUnreachableLink.ruleName = ruleName;
logUnreachableLink.ruleOptions = defaultRuleOptions;

module.exports = logUnreachableLink;
