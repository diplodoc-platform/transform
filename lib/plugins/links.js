const path = require('path');
const url = require('url');
const MarkdownIt = require('markdown-it');
const {
    isLocalUrl,
    isFileExists,
    getFileTokens,
    findBlockTokens,
    headingInfo,
    getSinglePageAnchorId,
} = require('../utils');
const {
    logEmptyLinkHref,
    logTitleRefLinkNotFound,
    logUnreachableLink,
} = require('../lintRules/pluginsRules');

const PAGE_LINK_REGEXP = /\.(md|ya?ml)$/i;

function defaultTransformLink(href) {
    const parsed = url.parse(href);

    return url.format({
        ...parsed,
        pathname: parsed.pathname.replace(PAGE_LINK_REGEXP, '.html'),
    });
}

function getTitleFromTokens(tokens) {
    let title = '';

    let i = 0;
    while (i < tokens.length) {
        const token = tokens[i];

        if (token.type === 'heading_open') {
            ({title = ''} = headingInfo(tokens, i));

            break;
        }

        i += 2;
    }

    return title;
}

const addTitle = ({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath}) => {
    const {log, lintOptions, disableLint} = opts;

    const id = hash && hash.slice(1);
    const fileTokens = getFileTokens(file, state, opts);
    const sourceTokens = id ? findBlockTokens(fileTokens, id) : fileTokens;
    const title = getTitleFromTokens(sourceTokens);

    if (title) {
        let textToken;

        if (isEmptyLink) {
            textToken = new state.Token('text', '', 0);
            tokens.splice(idx + 1, 0, textToken);
        } else {
            textToken = nextToken;
        }

        textToken.content = title;
    } else {
        logTitleRefLinkNotFound({
            lintOptions,
            commonOptions: {log, path: currentPath, href, disableLint},
        });
    }
};

// eslint-disable-next-line complexity
function processLink(state, tokens, idx, opts) {
    const {path: startPath, root, transformLink, notFoundCb, needSkipLinkFn, log, lintOptions, disableLint} = opts;
    const currentPath = state.env.path || startPath;
    const linkToken = tokens[idx];
    const nextToken = tokens[idx + 1];
    let href = linkToken.attrGet('href');

    if (!href) {
        logEmptyLinkHref({
            lintOptions,
            commonOptions: {log, path: startPath, disableLint},
        });
        return;
    }

    const {pathname, hash} = url.parse(href);
    let file;
    let fileExists;
    let isPageFile;

    if (!isLocalUrl(href)) {
        linkToken.attrSet('target', '_blank');
        linkToken.attrSet('rel', 'noreferrer noopener');
        return;
    }

    if (pathname) {
        file = path.resolve(path.parse(currentPath).dir, pathname);
        fileExists = isFileExists(file);
        isPageFile = PAGE_LINK_REGEXP.test(pathname);

        if (isPageFile && !fileExists) {
            let needShowError = true;
            if (needSkipLinkFn) {
                needShowError = !needSkipLinkFn(href);
            }

            if (notFoundCb && needShowError) {
                notFoundCb(file.replace(root, ''));
            }

            if (needShowError) {
                logUnreachableLink({
                    lintOptions,
                    commonOptions: {log, path: currentPath, href, disableLint},
                });
            }
        }
    } else if (hash) {
        file = startPath;
        fileExists = true;
        isPageFile = true;
    } else {
        return;
    }

    const isEmptyLink = nextToken.type === 'link_close';
    const isTitleRefLink = nextToken.type === 'text' && nextToken.content === '{#T}';
    if ((isEmptyLink || isTitleRefLink) && fileExists && isPageFile) {
        addTitle({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath});
    }

    let newPathname;
    if (currentPath !== startPath) {
        newPathname = path.relative(path.parse(startPath).dir, file);

        href = url.format({
            ...url.parse(href),
            pathname: newPathname,
        });
    }

    if (pathname || newPathname) {
        const transformer = transformLink || defaultTransformLink;
        linkToken.attrSet('href', transformer(href));
    }
}

function links(md, opts) {
    const plugin = (state) => {
        const tokens = state.tokens;
        let i = 0;

        state.md.disable(['includes', 'links'], true);

        while (i < tokens.length) {
            if (tokens[i].type === 'inline') {
                const childrenTokens = tokens[i].children;
                let j = 0;

                while (j < childrenTokens.length) {
                    const isLinkOpenToken = childrenTokens[j].type === 'link_open';
                    const tokenClass = childrenTokens[j].attrGet('class');

                    /*  Don't process anchor links */
                    const isYfmAnchor = tokenClass
                        ? tokenClass.includes('yfm-anchor')
                        : false;

                    if (isLinkOpenToken && !isYfmAnchor) {
                        processLink(state, childrenTokens, j, opts);
                    }

                    j++;
                }
            }

            i++;
        }

        state.md.enable(['includes', 'links'], true);
    };

    try {
        md.core.ruler.before('includes', 'links', plugin);
    } catch (e) {
        md.core.ruler.push('links', plugin);
    }
}

const replaceLinkHref = (input, href, newHref) => {
    /* Try not replace include syntax */
    return input.replace(`](${href})`, `](${newHref})`);
};

/* Replace the links to the markdown and yaml files if the singlePage option is passed in the options
*  Example: replace [Text](../../path/to/file.md#anchor) with [Text](#_path_to_file_anchor)
* */
links.collect = (input, options) => {
    const {root, path: startPath, singlePage} = options;

    if (!singlePage) {
        return;
    }

    let result = input;

    /* Syntax "{% include [Text](_includes/file.md) %}" is parsed as link. Need to ignore errors */
    const needSkipLinkFn = (href) => href.includes('_includes/');
    const transformLink = (href) => href;
    const md = new MarkdownIt()
        .use(links, {...options, transformLink, needSkipLinkFn});
    const tokens = md.parse(result, {});

    let i = 0;
    while (i < tokens.length) {
        if (tokens[i].type === 'inline') {
            const childrenTokens = tokens[i].children;
            let j = 0;

            while (j < childrenTokens.length) {
                const isLinkOpenToken = childrenTokens[j].type === 'link_open';

                if (!isLinkOpenToken) {
                    j++;
                    continue;
                }

                const linkToken = childrenTokens[j];
                const href = linkToken.attrGet('href');

                if (!href || !isLocalUrl(href) || href.includes('_includes/')) {
                    j++;
                    continue;
                }

                const {pathname, hash} = url.parse(href);
                if (pathname) {
                    const isPageFile = PAGE_LINK_REGEXP.test(pathname);
                    if (isPageFile) {
                        const newHref = getSinglePageAnchorId({root, currentPath: startPath, pathname, hash});
                        result = replaceLinkHref(result, href, newHref);
                    }
                } else if (hash) {
                    const newHref = getSinglePageAnchorId({root, currentPath: startPath, hash});
                    result = replaceLinkHref(result, href, newHref);
                }

                j++;
            }
        }

        i++;
    }

    // eslint-disable-next-line consistent-return
    return result;
};


module.exports = links;
