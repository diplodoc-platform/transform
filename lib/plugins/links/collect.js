const MarkdownIt = require('markdown-it');
const url = require('url');
const {
    isLocalUrl,
} = require('../../utils');
const {getSinglePageAnchorId} = require('../../utilsFS');
const index = require('./index');
const {PAGE_LINK_REGEXP} = require('./constants');

const replaceLinkHref = (input, href, newHref) => {
    /* Try not replace include syntax */
    return input.replace(`](${href})`, `](${newHref})`);
};

/* Replace the links to the markdown and yaml files if the singlePage option is passed in the options
*  Example: replace [Text](../../path/to/file.md#anchor) with [Text](#_path_to_file_anchor)
* */
const collect = (input, options) => {
    const {root, path: startPath, singlePage} = options;

    if (!singlePage) {
        return;
    }

    let result = input;

    /* Syntax "{% include [Text](_includes/file.md) %}" is parsed as link. Need to ignore errors */
    const needSkipLinkFn = (href) => href.includes('_includes/');
    const transformLink = (href) => href;
    const md = new MarkdownIt()
        .use(index, {...options, transformLink, needSkipLinkFn});
    const tokens = md.parse(result, {});

    let i = 0;
    while (i < tokens.length) {
        if (tokens[i].type === 'inline') {
            const childrenTokens = tokens[i].children;
            let j = 0;

            while (j < childrenTokens.length) {
                const isLinkOpenToken = childrenTokens[j].type === 'link_open';

                if (!isLinkOpenToken) {
                    j++;
                    continue;
                }

                const linkToken = childrenTokens[j];
                const href = linkToken.attrGet('href');

                if (!href || !isLocalUrl(href) || href.includes('_includes/')) {
                    j++;
                    continue;
                }

                const {pathname, hash} = url.parse(href);
                const isPageFile = PAGE_LINK_REGEXP.test(pathname);
                if (isPageFile) {
                    const newHref = getSinglePageAnchorId({root, currentPath: startPath, pathname, hash});
                    result = replaceLinkHref(result, href, newHref);
                }

                j++;
            }
        }

        i++;
    }

    // eslint-disable-next-line consistent-return
    return result;
};


module.exports = collect;
