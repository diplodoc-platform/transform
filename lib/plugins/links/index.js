const path = require('path');
const url = require('url');
const {bold} = require('chalk');
const {
    isLocalUrl,
    findBlockTokens,
    headingInfo,
} = require('../../utils');
const {
    isFileExists,
    getFileTokens,
} = require('../../utilsFS');
const {PAGE_LINK_REGEXP} = require('./constants');

function defaultTransformLink(href) {
    const parsed = url.parse(href);

    return url.format({
        ...parsed,
        pathname: parsed.pathname.replace(PAGE_LINK_REGEXP, '.html'),
    });
}

function getTitleFromTokens(tokens) {
    let title = '';

    let i = 0;
    while (i < tokens.length) {
        const token = tokens[i];

        if (token.type === 'heading_open') {
            ({title = ''} = headingInfo(tokens, i));

            break;
        }

        i += 2;
    }

    return title;
}

const addTitle = ({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath, log}) => {
    const id = hash && hash.slice(1);
    const fileTokens = getFileTokens(file, state, opts);
    const sourceTokens = id ? findBlockTokens(fileTokens, id) : fileTokens;
    const title = getTitleFromTokens(sourceTokens);

    if (title) {
        let textToken;

        if (isEmptyLink) {
            textToken = new state.Token('text', '', 0);
            tokens.splice(idx + 1, 0, textToken);
        } else {
            textToken = nextToken;
        }

        textToken.content = title;
    } else {
        log.warn(`Title not found: ${bold(href)} in ${bold(currentPath)}`);
    }
};

// eslint-disable-next-line complexity
function processLink(state, tokens, idx, opts) {
    const {path: startPath, root, transformLink, notFoundCb, needSkipLinkFn, log} = opts;
    const currentPath = state.env.path || startPath;
    const linkToken = tokens[idx];
    const nextToken = tokens[idx + 1];
    let href = linkToken.attrGet('href');

    if (!href) {
        log.error(`Empty link in ${bold(startPath)}`);
        return;
    }

    const {pathname, hash} = url.parse(href);
    let file;
    let fileExists;
    let isPageFile;

    if (!isLocalUrl(href)) {
        linkToken.attrSet('target', '_blank');
        linkToken.attrSet('rel', 'noreferrer noopener');
        return;
    }

    if (pathname) {
        file = path.resolve(path.parse(currentPath).dir, pathname);
        fileExists = isFileExists(file);
        isPageFile = PAGE_LINK_REGEXP.test(pathname);

        if (isPageFile && !fileExists) {
            let needShowError = true;
            if (needSkipLinkFn) {
                needShowError = !needSkipLinkFn(href);
            }

            if (notFoundCb && needShowError) {
                notFoundCb(file.replace(root, ''));
            }

            if (needShowError) {
                log.error(`Link is unreachable: ${bold(href)} in ${bold(currentPath)}`);
            }
        }
    } else if (hash) {
        file = startPath;
        fileExists = true;
        isPageFile = true;
    } else {
        return;
    }

    const isEmptyLink = nextToken.type === 'link_close';
    const isTitleRefLink = nextToken.type === 'text' && nextToken.content === '{#T}';
    if ((isEmptyLink || isTitleRefLink) && fileExists && isPageFile) {
        addTitle({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath, log});
    }

    let newPathname;
    if (currentPath !== startPath) {
        newPathname = path.relative(path.parse(startPath).dir, file);

        href = url.format({
            ...url.parse(href),
            pathname: newPathname,
        });
    }

    if (pathname || newPathname) {
        const transformer = transformLink || defaultTransformLink;
        linkToken.attrSet('href', transformer(href));
    }
}

function index(md, opts) {
    const plugin = (state) => {
        const tokens = state.tokens;
        let i = 0;

        state.md.disable(['includes', 'links'], true);

        while (i < tokens.length) {
            if (tokens[i].type === 'inline') {
                const childrenTokens = tokens[i].children;
                let j = 0;

                while (j < childrenTokens.length) {
                    const isLinkOpenToken = childrenTokens[j].type === 'link_open';
                    const tokenClass = childrenTokens[j].attrGet('class');

                    /*  Don't process anchor links */
                    const isYfmAnchor = tokenClass
                        ? tokenClass.includes('yfm-anchor')
                        : false;

                    if (isLinkOpenToken && !isYfmAnchor) {
                        processLink(state, childrenTokens, j, opts);
                    }

                    j++;
                }
            }

            i++;
        }

        state.md.enable(['includes', 'links'], true);
    };

    try {
        md.core.ruler.before('includes', 'links', plugin);
    } catch (e) {
        md.core.ruler.push('links', plugin);
    }
}

module.exports = index;
