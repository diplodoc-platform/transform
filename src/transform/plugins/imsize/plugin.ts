import type MarkdownIt from 'markdown-it';
import type ParserInline from 'markdown-it/lib/parser_inline';
import type Token from 'markdown-it/lib/token';

import {ImsizeAttr} from './const';
import {parseImageSize} from './helpers';

export type ImsizeOptions = {
    enableInlineStyling?: boolean;
};

export const imageWithSize = (md: MarkdownIt, opts?: ImsizeOptions): ParserInline.RuleInline => {
    // eslint-disable-next-line complexity
    return (state, silent) => {
        if (state.src.charCodeAt(state.pos) !== 0x21 /* ! */) {
            return false;
        }

        if (state.src.charCodeAt(state.pos + 1) !== 0x5b /* [ */) {
            return false;
        }

        const labelStart = state.pos + 2;
        const labelEnd = md.helpers.parseLinkLabel(state, state.pos + 1, false);

        // parser failed to find ']', so it's not a valid link
        if (labelEnd < 0) {
            return false;
        }

        let href = '';
        let title = '';
        let width = '';
        let height = '';

        const oldPos = state.pos;
        const max = state.posMax;

        let pos = labelEnd + 1;
        if (pos < max && state.src.charCodeAt(pos) === 0x28 /* ( */) {
            //
            // Inline link
            //

            // [link](  <href>  "title"  )
            //        ^^ skipping these spaces
            pos++;
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }
            if (pos >= max) {
                return false;
            }

            // [link](  <href>  "title"  )
            //          ^^^^^^ parsing link destination
            let start = pos;
            const destResult = md.helpers.parseLinkDestination(state.src, pos, state.posMax);
            if (destResult.ok) {
                href = state.md.normalizeLink(destResult.str);
                if (state.md.validateLink(href)) {
                    pos = destResult.pos;
                } else {
                    href = '';
                }
            }

            // [link](  <href>  "title"  )
            //                ^^ skipping these spaces
            start = pos;
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }

            // [link](  <href>  "title"  )
            //                  ^^^^^^^ parsing link title
            const titleResult = md.helpers.parseLinkTitle(state.src, pos, state.posMax);
            if (pos < max && start !== pos && titleResult.ok) {
                title = titleResult.str;
                pos = titleResult.pos;

                // [link](  <href>  "title"  )
                //                         ^^ skipping these spaces
                for (; pos < max; pos++) {
                    const code = state.src.charCodeAt(pos);
                    if (code !== 0x20 && code !== 0x0a) {
                        break;
                    }
                }
            } else {
                title = '';
            }

            // [link](  <href>  "title" =WxH  )
            //                          ^^^^ parsing image size
            if (pos - 1 >= 0) {
                const code = state.src.charCodeAt(pos - 1);

                // there must be at least one white spaces
                // between previous field and the size
                if (code === 0x20) {
                    const sizeResult = parseImageSize(state.src, pos, state.posMax);
                    if (sizeResult.ok) {
                        width = sizeResult.width;
                        height = sizeResult.height;
                        pos = sizeResult.pos;

                        // [link](  <href>  "title" =WxH  )
                        //                              ^^ skipping these spaces
                        for (; pos < max; pos++) {
                            const code = state.src.charCodeAt(pos);
                            // eslint-disable-next-line max-depth
                            if (code !== 0x20 && code !== 0x0a) {
                                break;
                            }
                        }
                    }
                }
            }

            if (pos >= max || state.src.charCodeAt(pos) !== 0x29 /* ) */) {
                state.pos = oldPos;
                return false;
            }
            pos++;
        } else {
            //
            // Link reference
            //
            if (typeof state.env.references === 'undefined') {
                return false;
            }

            // [foo]  [bar]
            //      ^^ optional whitespace (can include newlines)
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }

            let label = '';

            if (pos < max && state.src.charCodeAt(pos) === 0x5b /* [ */) {
                const start = pos + 1;
                pos = md.helpers.parseLinkLabel(state, pos);
                if (pos >= 0) {
                    label = state.src.slice(start, pos++);
                } else {
                    pos = labelEnd + 1;
                }
            } else {
                pos = labelEnd + 1;
            }

            // covers label === '' and label === undefined
            // (collapsed reference link and shortcut reference link respectively)
            if (!label) {
                label = state.src.slice(labelStart, labelEnd);
            }

            const ref = state.env.references[md.utils.normalizeReference(label)];
            if (!ref) {
                state.pos = oldPos;
                return false;
            }
            href = ref.href;
            title = ref.title;
        }

        //
        // We found the end of the link, and know for a fact it's a valid link;
        // so all that's left to do is to call tokenizer.
        //
        if (!silent) {
            state.pos = labelStart;
            state.posMax = labelEnd;

            const tokens: Token[] = [];
            const newState = new state.md.inline.State(
                state.src.slice(labelStart, labelEnd),
                state.md,
                state.env,
                tokens,
            );
            newState.md.inline.tokenize(newState);

            const token = state.push('image', 'img', 0);
            token.children = tokens;
            token.attrs = [
                [ImsizeAttr.Src, href],
                [ImsizeAttr.Alt, ''],
            ];

            if (title) {
                token.attrs.push([ImsizeAttr.Title, title]);
            }

            if (width !== '') {
                token.attrs.push([ImsizeAttr.Width, width]);
            }

            if (height !== '') {
                token.attrs.push([ImsizeAttr.Height, height]);
            }

            if (opts?.enableInlineStyling) {
                let style: string | undefined = '';

                const widthWithPercent = width.includes('%');
                const heightWithPercent = height.includes('%');

                if (width !== '') {
                    const widthString = widthWithPercent ? width : `${width}px`;
                    style += `width: ${widthString};`;
                }

                if (height !== '') {
                    if (width !== '' && !heightWithPercent && !widthWithPercent) {
                        style += `aspect-ratio: ${width} / ${height};height: auto;`;
                        state.env.additionalOptionsCssWhiteList ??= {};
                        state.env.additionalOptionsCssWhiteList['aspect-ratio'] = true;
                    } else {
                        const heightString = heightWithPercent ? height : `${height}px`;
                        style += `height: ${heightString};`;
                    }
                }

                if (style) {
                    token.attrs.push([ImsizeAttr.Style, style]);
                }
            }
        }

        state.pos = pos;
        state.posMax = max;
        return true;
    };
};
