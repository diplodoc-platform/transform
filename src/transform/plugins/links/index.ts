import url from 'url';
import {bold} from 'chalk';
import {
    PAGE_LINK_REGEXP,
    defaultTransformLink,
    findBlockTokens,
    getHrefTokenAttr,
    headingInfo,
    isLocalUrl,
} from '../../utils';
import {getFileTokens, isFileExists} from '../../utilsFS';
import Token from 'markdown-it/lib/token';
import {Logger} from 'src/transform/log';
import {MarkdownItPluginCb, MarkdownItPluginOpts} from '../typings';
import path, {isAbsolute, parse, relative, resolve} from 'path';
import {StateCore} from 'src/transform/typings';

function getTitleFromTokens(tokens: Token[]) {
    let title = '';

    let i = 0;
    while (i < tokens.length) {
        const token = tokens[i];

        if (token.type === 'heading_open') {
            ({title = ''} = headingInfo(tokens, i));

            break;
        }

        i += 2;
    }

    return title;
}

type Options = {
    hash: string | null;
    file: string;
    state: StateCore;
    opts: object;
    isEmptyLink: boolean;
    tokens: Token[];
    idx: number;
    nextToken: Token;
    href: string;
    currentPath: string;
    log: Logger;
};

const addTitle = (options: Options) => {
    const {hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath, log} =
        options;

    const id = hash && hash.slice(1);
    const fileTokens = getFileTokens(file, state, {
        ...opts,
        disableLint: true,
        disableTitleRefSubstitution: true,
        disableCircularError: true,
        inheritVars: false,
    });
    const sourceTokens = id ? findBlockTokens(fileTokens, id) : fileTokens;
    const title = getTitleFromTokens(sourceTokens);

    if (title) {
        let textToken;

        if (isEmptyLink) {
            textToken = new state.Token('text', '', 0);
            tokens.splice(idx + 1, 0, textToken);
        } else {
            textToken = nextToken;
        }

        textToken.content = title;
    } else {
        log.warn(`Title not found: ${bold(href)} in ${bold(currentPath)}`);
    }
};

interface ProcOpts extends MarkdownItPluginOpts {
    transformLink: (v: string) => string;
    notFoundCb: (v: string) => void;
    needSkipLinkFn: (v: string) => boolean;
    getPublicPath: (options: ProcOpts, v?: string) => string;
}

function getDefaultPublicPath(
    {
        file,
        path,
    }: {
        file?: string;
        path?: string;
    },
    input?: string | null,
) {
    return relative(parse(path || '').dir, input || file || '');
}

// eslint-disable-next-line complexity
function processLink(state: StateCore, tokens: Token[], idx: number, opts: ProcOpts) {
    const {
        path: startPath,
        root,
        transformLink,
        notFoundCb,
        needSkipLinkFn,
        log,
        getPublicPath = getDefaultPublicPath,
    } = opts;

    const currentPath = state.env.path || startPath;
    const linkToken = tokens[idx];
    const nextToken = tokens[idx + 1];

    let href = getHrefTokenAttr(linkToken);

    if (!href) {
        log.error(`Empty link in ${bold(startPath)}`);
        return;
    }

    const {pathname, hash} = url.parse(href);
    let file;
    let fileExists;
    let isPageFile;

    if (!isLocalUrl(href)) {
        linkToken.attrSet('target', '_blank');
        linkToken.attrSet('rel', 'noreferrer noopener');
        return;
    }

    if (pathname) {
        file = resolve(path.parse(currentPath).dir, pathname);
        fileExists = isFileExists(file);
        isPageFile = PAGE_LINK_REGEXP.test(pathname);

        if (isPageFile && !fileExists) {
            let needShowError = true;
            if (needSkipLinkFn) {
                needShowError = !needSkipLinkFn(href);
            }

            if (notFoundCb && needShowError) {
                notFoundCb(file.replace(root, ''));
            }

            if (needShowError) {
                log.error(`Link is unreachable: ${bold(href)} in ${bold(currentPath)}`);
            }
        }
    } else if (hash) {
        file = startPath;
        fileExists = true;
        isPageFile = true;
    } else {
        return;
    }

    const isEmptyLink = nextToken.type === 'link_close';
    const isTitleRefLink = nextToken.type === 'text' && nextToken.content === '{#T}';
    if (
        (isEmptyLink || isTitleRefLink) &&
        fileExists &&
        isPageFile &&
        !state.env.disableTitleRefSubstitution
    ) {
        addTitle({
            hash,
            file,
            state,
            opts,
            isEmptyLink,
            tokens,
            idx,
            nextToken,
            href,
            currentPath,
            log,
        });
    }

    let newPathname = '';

    if (!isAbsolute(href) && !href.includes('//')) {
        newPathname = getPublicPath(opts, file);

        href = url.format({
            ...url.parse(href),
            pathname: newPathname,
        });
    }

    if (pathname || newPathname) {
        const transformer = transformLink || defaultTransformLink;
        linkToken.attrSet('href', transformer(href));
    }
}

const index: MarkdownItPluginCb<ProcOpts & Options> = (md, opts) => {
    const plugin = (state: StateCore) => {
        const tokens = state.tokens;
        let i = 0;

        while (i < tokens.length) {
            if (tokens[i].type === 'inline') {
                const childrenTokens = tokens[i].children || [];
                let j = 0;

                while (j < childrenTokens.length) {
                    const isLinkOpenToken = childrenTokens[j].type === 'link_open';
                    const tokenClass = childrenTokens[j].attrGet('class');

                    /*  Don't process anchor links */
                    const isYfmAnchor = tokenClass ? tokenClass.includes('yfm-anchor') : false;

                    if (isLinkOpenToken && !isYfmAnchor) {
                        processLink(state, childrenTokens, j, opts);
                    }

                    j++;
                }
            }

            i++;
        }
    };

    try {
        md.core.ruler.before('includes', 'links', plugin);
    } catch (e) {
        md.core.ruler.push('links', plugin);
    }
};

export = index;
