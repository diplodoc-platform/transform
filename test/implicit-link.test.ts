import initMarkdown from '../src/transform/md';
import plugin from '../src/transform/plugins/block-anchor';
import implicitLink from '../src/transform/plugins/implicit-link';
import links from '../src/transform/plugins/links';


const {parse, compile} = initMarkdown({plugins: [plugin, links, implicitLink]});

describe('implicit-link', function () {
    it('renders implicit-link', () => {
        const input = 'https://google.com';
        const actual = compile(parse(input));
        expect(actual).toMatchSnapshot();
    });
});
u