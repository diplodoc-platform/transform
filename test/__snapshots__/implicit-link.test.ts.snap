// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`implicit-link renders implicit-link 1`] = `"<a href=\\"https://google.com\\">https://google.com</a>"`;
